vortex-serde/src/layout/mod.rs

```diff
@@ -0,0 +1,507 @@
+    layout_refs: HashMap<LayoutId, LayoutRef>,
+}
+
+impl LayoutContext {
+    pub fn new(layout_refs: HashMap<LayoutId, LayoutRef>) -> Self {
+        Self { layout_refs }
+    }
+
+    pub fn lookup_layout_def(&self, id: &LayoutId) -> Option<LayoutRef> {
+        self.layout_refs.get(id).cloned()
+    }
+}
+
+impl Default for LayoutContext {
+    fn default() -> Self {
+        Self::new(
+            [
+                &ColumnLayoutDefinition as LayoutRef,
+                &ChunkedLayoutDefinition,
+            ]
+            .into_iter()
+            .map(|l| (l.id(), l))
+            .collect(),
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LayoutReader {
+    ctx: Arc<Context>,
+    layout_ctx: Arc<LayoutContext>,
+}
+
+impl LayoutReader {
+    pub fn new(ctx: Arc<Context>, layout_ctx: Arc<LayoutContext>) -> Self {
+        Self { ctx, layout_ctx }
+    }
+
+    pub fn read(
+        &self,
+        fb_bytes: Bytes,
+        fb_loc: usize,
+        scan: Scan,
+        message_cache: RelativeMessageCache,
+    ) -> VortexResult<Box<dyn Layout>> {
+        let fb_layout = unsafe {
+            let tab = flatbuffers::Table::new(&fb_bytes, fb_loc);
+            fb::Layout::init_from_table(tab)
+        };
+
+        match fb_layout.layout_type() {
+            LayoutVariant::FlatLayout => {
+                let flat_layout = fb_layout.layout_as_flat_layout().expect("must be flat");
+                Ok(Box::new(FlatLayout::new(
+                    flat_layout.begin(),
+                    flat_layout.end(),
+                    self.ctx.clone(),
+                    message_cache,
+                )))
+            }
+            LayoutVariant::NestedLayout => {
+                let nested_layout = fb_layout
+                    .layout_as_nested_layout()
+                    .expect("must be nested layout");
+                Ok(self
+                    .layout_ctx
+                    .lookup_layout_def(&LayoutId(nested_layout.encoding()))
+                    .ok_or_else(|| {
+                        vortex_err!("Unknown layout definition {}", nested_layout.encoding())
+                    })?
+                    .layout(fb_bytes, fb_loc, scan, self.clone(), message_cache))
+            }
+            _ => unreachable!("Unknown flatbuffer layout"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Scan {
+    indices: Option<Array>,
+    projection: Projection,
+    filter: Option<RowFilter>,
+    batch_size: usize,
+}
+
+/// Unique identifier for a message within a layout
+pub type MessageId = u16;
+
+#[derive(Debug)]
+pub enum ReadResult {
+    GetMsgs(
+        Vec<(Vec<MessageId>, ByteRange)>,
+        Vec<(Vec<MessageId>, ByteRange)>,
+    ),
+    Batch(Array),
+}
+
+pub enum PlanResult {
+    GetMsg(Vec<(Vec<MessageId>, ByteRange)>),
+    Batch(Vec<MessageId>, Array),
+}
+
+#[derive(Debug)]
+pub struct MessagesCache {
+    cache: HashMap<Vec<MessageId>, Bytes>,
+}
+
+impl Default for MessagesCache {
```

Comment:
this implementation can just be derived IMO, and then you can call `Default::default()` in `MessagesCache::new`